### ISSUE TITLE
Blazor windows authentication authorization issues with ClaimsTransformation

### ISSUE THREAD

**Issue by FrancisValade** [aspnet/AspNetCore#11633]
1. Create a new Blazor server-side app with the preview 6-012264 using the blazor extension in Visual Studio and add windows authentication.
2. Add the following class in the app with the following code: 

```
    public class ClaimsTransformerTest: IClaimsTransformation
    {
        public async Task<ClaimsPrincipal> TransformAsync(ClaimsPrincipal principal)
        {
            ClaimsIdentity identite = (ClaimsIdentity)principal.Identity;
            if (identite.IsAuthenticated && identite is WindowsIdentity)
            {
                    identite.AddClaim(new Claim("Admin", "Role1"));
            }
            await Task.Delay(10);
            return new ClaimsPrincipal(identite);
        }
    }
```
And add the following lines in the ConfigureServices method in the startup class:
```
services.AddAuthentication(IISDefaults.AuthenticationScheme);
services.AddAuthorization(options =>
{
    options.AddPolicy("AdminRole1", policy =>
        policy.RequireClaim("Admin", "Role1"));
});
services.AddScoped<IClaimsTransformation, ClaimsTransformerTest>(); 
```
And in the Configure method add these lines: 
```
 app.UseAuthentication();
 app.UseAuthorization();
```
And finally add the Authorize attribute to the FetchData.razor or any page that you want to access:
```
@using Microsoft.AspNetCore.Authorization;
@attribute [Authorize(Policy = "AdminRole1")]
```
3. Run the app in debug and try to access the page in which you added the authorization policy
4. I get the following error: 
```
System.ObjectDisposedException: Safe handle has been closed.
Object name: 'SafeHandle'.
   at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
   at System.StubHelpers.StubHelpers.SafeHandleAddRef(SafeHandle pHandle, Boolean& success)
   at Interop.Advapi32.GetTokenInformation(SafeAccessTokenHandle TokenHandle, UInt32 TokenInformationClass, SafeLocalAllocHandle TokenInformation, UInt32 TokenInformationLength, UInt32& ReturnLength)
   at System.Security.Principal.WindowsIdentity.GetTokenInformation(SafeAccessTokenHandle tokenHandle, TokenInformationClass tokenInformationClass, Boolean nullOnInvalidParam)
   at System.Security.Principal.WindowsIdentity.get_User()
   at System.Security.Principal.WindowsIdentity.<GetName>b__51_0()
   at System.Security.Principal.WindowsIdentity.<>c__DisplayClass67_0.<RunImpersonatedInternal>b__0(Object <p0>)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Security.Principal.WindowsIdentity.RunImpersonatedInternal(SafeAccessTokenHandle token, Action action)
   at System.Security.Principal.WindowsIdentity.RunImpersonated(SafeAccessTokenHandle safeAccessTokenHandle, Action action)
   at System.Security.Principal.WindowsIdentity.GetName()
   at System.Security.Principal.WindowsIdentity.get_Name()
   at System.Security.Principal.WindowsIdentity.InitializeClaims()
   at System.Security.Principal.WindowsIdentity.get_Claims()+MoveNext()
   at System.Security.Claims.ClaimsPrincipal.get_Claims()+MoveNext()
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at Microsoft.AspNetCore.Authorization.Infrastructure.ClaimsAuthorizationRequirement.HandleRequirementAsync(AuthorizationHandlerContext context, ClaimsAuthorizationRequirement requirement)
   at Microsoft.AspNetCore.Authorization.AuthorizationHandler`1.HandleAsync(AuthorizationHandlerContext context)
   at Microsoft.AspNetCore.Authorization.Infrastructure.PassThroughAuthorizationHandler.HandleAsync(AuthorizationHandlerContext context)
   at Microsoft.AspNetCore.Authorization.DefaultAuthorizationService.AuthorizeAsync(ClaimsPrincipal user, Object resource, IEnumerable`1 requirements)
   at Microsoft.AspNetCore.Components.AuthorizeViewCore.IsAuthorizedAsync(ClaimsPrincipal user)
   at Microsoft.AspNetCore.Components.AuthorizeViewCore.OnParametersSetAsync()
   at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion(Task task)
   at Microsoft.AspNetCore.Components.ComponentBase.RunInitAndSetParametersAsync()
Microsoft.AspNetCore.Components.Server.ComponentHub: Warning: Unhandled Server-Side exception

System.ObjectDisposedException: Safe handle has been closed.
Object name: 'SafeHandle'.
   at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
   at System.StubHelpers.StubHelpers.SafeHandleAddRef(SafeHandle pHandle, Boolean& success)
   at Interop.Advapi32.GetTokenInformation(SafeAccessTokenHandle TokenHandle, UInt32 TokenInformationClass, SafeLocalAllocHandle TokenInformation, UInt32 TokenInformationLength, UInt32& ReturnLength)
   at System.Security.Principal.WindowsIdentity.GetTokenInformation(SafeAccessTokenHandle tokenHandle, TokenInformationClass tokenInformationClass, Boolean nullOnInvalidParam)
   at System.Security.Principal.WindowsIdentity.get_User()
   at System.Security.Principal.WindowsIdentity.<GetName>b__51_0()
   at System.Security.Principal.WindowsIdentity.<>c__DisplayClass67_0.<RunImpersonatedInternal>b__0(Object <p0>)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Security.Principal.WindowsIdentity.RunImpersonatedInternal(SafeAccessTokenHandle token, Action action)
   at System.Security.Principal.WindowsIdentity.RunImpersonated(SafeAccessTokenHandle safeAccessTokenHandle, Action action)
   at System.Security.Principal.WindowsIdentity.GetName()
   at System.Security.Principal.WindowsIdentity.get_Name()
   at System.Security.Principal.WindowsIdentity.InitializeClaims()
   at System.Security.Principal.WindowsIdentity.get_Claims()+MoveNext()
   at System.Security.Claims.ClaimsPrincipal.get_Claims()+MoveNext()
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at Microsoft.AspNetCore.Authorization.Infrastructure.ClaimsAuthorizationRequirement.HandleRequirementAsync(AuthorizationHandlerContext context, ClaimsAuthorizationRequirement requirement)
   at Microsoft.AspNetCore.Authorization.AuthorizationHandler`1.HandleAsync(AuthorizationHandlerContext context)
   at Microsoft.AspNetCore.Authorization.Infrastructure.PassThroughAuthorizationHandler.HandleAsync(AuthorizationHandlerContext context)
   at Microsoft.AspNetCore.Authorization.DefaultAuthorizationService.AuthorizeAsync(ClaimsPrincipal user, Object resource, IEnumerable`1 requirements)
   at Microsoft.AspNetCore.Components.AuthorizeViewCore.IsAuthorizedAsync(ClaimsPrincipal user)
   at Microsoft.AspNetCore.Components.AuthorizeViewCore.OnParametersSetAsync()
   at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion(Task task)
   at Microsoft.AspNetCore.Components.ComponentBase.RunInitAndSetParametersAsync()
Microsoft.AspNetCore.Routing.EndpointMiddleware: Information: Executed endpoint '/_blazor'
Microsoft.AspNetCore.Hosting.Diagnostics: Information: Request finished in 1487.5535ms 101 
```

**Comment by SteveSandersonMS** [comment]
Looks similar to some earlier issues with Windows authentication in non-Blazor cases:

 * https://github.com/aspnet/Logging/issues/543
 * https://github.com/aspnet/HttpSysServer/issues/426
 * https://github.com/aspnet/SignalR/issues/2948

I see that https://github.com/aspnet/SignalR/issues/2948 was fixed by @BrennanConroy making a change to clone the `ClaimsPrincipal` at some point in the flow.

@BrennanConroy, do you still remember what was going on with that SignalR issue? Do you know if we need to do something equivalent in Blazor? In Blazor we're holding a reference to the `User` that comes from `HubConnectionContext` - is there some reason why we can't hold and use that for the duration of the connection? When and why does it have to be cloned?

**Comment by blowdart** [comment]
I don't think you can change a Windows identity like this, because a windows identity is somewhat special, it's not a normal claims identity at all, it just maps to one via magic, and it will get its claims from AD.

Take the claims out of the windows identity, put them into a new ClaimsIdentity via the IEnumerable<Claim> constructor and then build a new principal off that.

**Comment by FrancisValade** [comment]
> I don't think you can change a Windows identity like this, because a windows identity is somewhat special, it's not a normal claims identity at all, it just maps to one via magic, and it will get its claims from AD.
> 
> Take the claims out of the windows identity, put them into a new ClaimsIdentity via the IEnumerable constructor and then build a new principal off that.

I tried what you suggested and the problem is gone, thanks!

**Comment by SteveSandersonMS** [comment]
Thanks for the info, @blowdart. Closing as "working as expected" then.

**Comment by BrennanConroy** [comment]
>In Blazor we're holding a reference to the User that comes from HubConnectionContext - is there some reason why we can't hold and use that for the duration of the connection?

As long as it's only for the duration of the connection you're good. SignalR will handle any silliness like the windows auth stuff.

**Comment by SteveSandersonMS** [comment]
@BrennanConroy

> As long as it's only for the duration of the connection you're good

Well, that's not completely the case. If the SignalR connection drops, we move the circuit into a "pending shutdown" state where it stays for a while, as we wait to see if the user reconnects. It's possible that some server-side code chooses to run while the circuit is in "pending shutdown" and such user code could interact with the `User` object.

What are the extra guarantees that SignalR offers while the connection is open that it stops doing later? How is SignalR actually affecting the `ClaimsPrincipal` object for some other code that is holding a reference to that object (and not just using `HubConnectionContext.User`)?

**Comment by BrennanConroy** [comment]
Well the `User` object isn't disposed while the SignalR connection is around. We don't explicitly dispose it, but it could be disposed by the server when the connection is gone.

@Tratcher I'm assuming it's a bad idea to hold onto the `User` from `HttpContext` after the underlying server connection is gone?

**Comment by Tratcher** [comment]
Right, it's disposed at the end of each request. Clone it if you need to hold onto it.

**Comment by AJB-Sharp** [comment]
I'm getting the following using Windows Authentication:
'user.Identity.Name' threw an exception of type 'System.ObjectDisposedException'
You suggests that I should clone the claims, which I do not know where or how to do. Please can you help.

**Comment by SteveSandersonMS** [comment]
This is an issue that will be fixed in the 3.0 Preview 8 release (the code is already in `master`). Until then there isn't an easy fix, though technically it's possible to implement your own `AuthenticationStateProvider` that does whatever you want.

**Comment by AJB-Sharp** [comment]
Thanks Steve
